### PR TITLE
DrivAerML: 4L/320d + WD=1e-2 + T_max=20 compound regularization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
WD=1e-2 and shorter T_max (20 vs 30) both helped on AirfRANS. This experiment tests both together on DrivAerML at the proven 4L/320d architecture. On AirfRANS, WD=1e-2 contributed a -6.8% improvement. T_max=20 on DrivAerML is already in-flight (ymir). This tests the compound effect of both simultaneously.

## Instructions

Run the following command in `target/icml2026`:

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 20 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group drivaerml-4L320d-compound
```

Key changes vs baseline:
- `--cosine-t-max 20` (was 30) — shorter cycles, more frequent restarts
- `--weight-decay 1e-2` (was none) — regularization from AirfRANS winning config
- `--grad-clip 1.5` — AirfRANS best gc value

Report `val_primary/surface_rel_l2_pct` at the best epoch. Also note the epoch at which the best was achieved and whether the model was still improving at the end of training.

## Baseline

- **val_primary/surface_rel_l2_pct:** 5.027% (epoch 257, 180-min budget, still improving)
- **test_primary/surface_rel_l2_pct:** 6.244%
- **Best PR:** #2648 (zoro — 4L/320d+T_max=30, lr=5e-4, 257 epochs)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 320 --model-heads 5 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

**CRITICAL:** Must pass `--batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200` — do not omit these.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2783 because ADVISOR: Closing — WD=1e-2 catastrophically diverges on DrivAerML (ray #2790 confirmed). Also architecture superseded by 4L/512d baseline..

**Comment by tcapelle** [comment]
STUDENT eren:

## Results

**Hypothesis:** WD=1e-2 + T_max=20 + gc=1.5 compound regularization on DrivAerML 4L/320d

**Verdict: DID NOT BEAT BASELINE — catastrophic divergence at epoch 103**

| Metric | Baseline (PR #2648) | This experiment | Delta |
|--------|-------------------|-----------------|-------|
| val_primary/surface_rel_l2_pct | 5.027% (ep 257) | 7.640% (ep 96) | +52% worse |

### Key metrics at best epoch (96)
- **val_primary/surface_rel_l2_pct:** 7.640%
- **Loss:** 0.0151
- **Grad norm mean:** 0.432
- **Grad clip events:** 0/394
- **Peak VRAM:** ~4.2 GB

### Command used
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 20 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-4L320d-compound \
  --agent eren --wandb-name "eren/drivaerml-4L320d-wd1e2-tmax20"
```

### W&B run
- **Run ID:** cagq6d49
- **URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/cagq6d49

### What happened

**Phase 1 — Strong convergence (epochs 1-101):** The compound config showed excellent initial convergence. The model reached 7.64% by epoch 96, with grad norms declining from 6.5 to 0.4 and grad clip events dropping from 337/394 to 0/394. The cosine trough envelope steadily deepened: 26.4% → 20.5% → 14.4% → 10.6% → 9.0% → 8.1% → 7.6%. At matched wall-clock, the trajectory was comparable to the baseline's early epochs.

**Phase 2 — First catastrophic explosion (epochs 103-125):** At epoch 102, grad norms suddenly jumped from 0.56 to 36.6 (epoch 103), then escalated to 62.9 → 103.6 → 175 → 217 → 274. Val metric spiked from 8.3% to 66.5%. The model briefly recovered at epoch 125 (grad norm collapsed from 122 to 1.58, val reached 23.6%) and was converging again.

**Phase 3 — Second explosion (epoch 130+):** At epoch 130, grad norms exploded again to 1284, then 7300 → 488 billion → Infinity. Model weights went to NaN. Unrecoverable.

### Analysis

The WD=1e-2 + gc=1.5 combination is **fundamentally unstable for long DrivAerML training**. The pattern suggests that weight decay accumulates a destabilizing effect over many epochs — the model performs well initially but inevitably hits a catastrophic gradient explosion. The gc=1.5 clip cannot contain the instability once it triggers. The brief recovery at epoch 125 suggests the model can temporarily restabilize, but the underlying instability recurs.

This independently confirms ray's findings in PR #2790 — WD=1e-2 diverges on DrivAerML.

### Suggested follow-ups
- WD=1e-3 or 5e-4 may provide regularization benefits without triggering instability
- Tighter grad clip (e.g. gc=1.0) combined with lower WD might prevent explosions
- The strong pre-explosion trajectory (7.64% at epoch 96 vs baseline 5.027% at epoch 257) suggests the compound config converges faster per-epoch — if the instability can be prevented, this approach could beat the baseline
